### DISCUSSION TITLE
Scaffold of some benzene-related molecules is not a benzene ring.

### DISCUSSION THREAD

**Discussion by zhikaili** [rdkit/rdkit#4288] [Q&A]
Hi,

I am using MurckoDecompose() to get scaffolds. I find that, for some molecules that contain a benzene ring, the output scaffold molecule object is however not a benzene ring. Two examples are shown below:

1.
![image](https://user-images.githubusercontent.com/76140761/123746141-d05f1480-d8e3-11eb-8392-2e3fb9707a92.png)

2.
![image](https://user-images.githubusercontent.com/76140761/123746188-dead3080-d8e3-11eb-9ba7-02318982d06f.png)

My question is:
1. Is this the expected behavior (chemically correct)?
2. If so, could you please briefly explain the reason? Thank you!
3. Why I can get 'c1ccccc1', which I believe is the smiles string of benzene, by feeding this output molecule object to Chem.MolToSmiles()?

Thank you very much!

**Comment by ptosco** [comment]
@zhikaili It looks like the Murcko decomposition code is not updating the implicit H counts after cutting the bonds from the scaffold to the substituents (see arrows below):
```
scaffold_mol.Debug()
RDKit INFO: Atoms:
RDKit INFO: 	0 6 C chg: 0  deg: 2 exp: 4 imp: 0 hyb: 3 arom?: 1 chi: 0  <====
RDKit INFO: 	1 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	2 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	3 6 C chg: 0  deg: 2 exp: 4 imp: 0 hyb: 3 arom?: 1 chi: 0 <====
RDKit INFO: 	4 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	5 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: Bonds:
RDKit INFO: 	0 0->1 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	1 1->2 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	2 2->3 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	3 3->4 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	4 4->5 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	5 5->0 order: 12 conj?: 1 aromatic?: 1
```
That looks incorrect to me, so I will file a bug; meanwhile, you can fix the problem by sanitizing the molecule:
```
Chem.SanitizeMol(scaffold_mol)
rdkit.Chem.rdmolops.SanitizeFlags.SANITIZE_NONE
scaffold_mol
```
![image](https://user-images.githubusercontent.com/5244385/123756972-c5918900-d8bd-11eb-89a1-a1f61e6c38ea.png)
```
scaffold_mol.Debug()
RDKit INFO: Atoms:
RDKit INFO: 	0 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0 <====
RDKit INFO: 	1 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	2 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	3 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0 <====
RDKit INFO: 	4 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: 	5 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
RDKit INFO: Bonds:
RDKit INFO: 	0 0->1 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	1 1->2 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	2 2->3 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	3 3->4 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	4 4->5 order: 12 conj?: 1 aromatic?: 1
RDKit INFO: 	5 5->0 order: 12 conj?: 1 aromatic?: 1
```

**Reply by zhikaili** [reply]
Hi, thank you for your reply!

**Reply by greglandrum** [reply]
Sorry I am just noticing this now: this is by design (and I commented on @ptosco's bug report to that effect).
Most of the RDKit code which manipulates molecules does not re-sanitize the molecule. This allows you to chain multiple operations together without paying the performance cost of constant re-sanitization.

As @ptosco points out: you can do a complete re-sanitization once you're done modifying the molecule.
